cpp/src/reader_c.cpp

```diff
@@ -0,0 +1,379 @@
+      properties->properties[i].key = nullptr;
+      properties->properties[i].value = nullptr;
+
+      if (keys[i]) {
+        size_t key_len = strlen(keys[i]) + 1;
+        properties->properties[i].key = static_cast<char*>(malloc(key_len));
+        if (properties->properties[i].key) {
+          strcpy(properties->properties[i].key, keys[i]);
+        }
+      }
+
+      if (values[i]) {
+        size_t value_len = strlen(values[i]) + 1;
+        properties->properties[i].value = static_cast<char*>(malloc(value_len));
+        if (properties->properties[i].value) {
+          strcpy(properties->properties[i].value, values[i]);
+        }
+      }
+    }
+
+    properties->count = count;
+  } catch (...) {
+    read_properties_free(properties);
+  }
+}
+
+const char* read_properties_get(const ::ReadProperties* properties, const char* key) {
+  if (!properties || !properties->properties || !key)
+    return nullptr;
+
+  for (size_t i = 0; i < properties->count; ++i) {
+    if (properties->properties[i].key && strcmp(properties->properties[i].key, key) == 0) {
+      return properties->properties[i].value;
+    }
+  }
+
+  return nullptr;
+}
+
+void read_properties_free(::ReadProperties* properties) {
+  if (!properties)
+    return;
+
+  if (properties->properties) {
+    for (size_t i = 0; i < properties->count; ++i) {
+      free(properties->properties[i].key);
+      free(properties->properties[i].value);
+    }
+    free(properties->properties);
+  }
+
+  properties->properties = nullptr;
+  properties->count = 0;
+}
+
+// ==================== ChunkReader C Implementation ====================
+
+void get_chunk_indices(ChunkReaderHandle reader,
+                       const int64_t* row_indices,
+                       size_t num_indices,
+                       int64_t** chunk_indices,
+                       size_t* num_chunk_indices) {
+  if (!reader || !row_indices || num_indices == 0 || !chunk_indices || !num_chunk_indices) {
+    if (chunk_indices)
+      *chunk_indices = nullptr;
+    if (num_chunk_indices)
+      *num_chunk_indices = 0;
+    return;
+  }
+
+  try {
+    auto* cpp_reader = static_cast<ChunkReader*>(reader);
+    std::vector<int64_t> input_indices(row_indices, row_indices + num_indices);
+
+    auto result = cpp_reader->get_chunk_indices(input_indices);
+    if (!result.ok()) {
+      *chunk_indices = nullptr;
+      *num_chunk_indices = 0;
+      return;
+    }
+
+    const auto& output_indices = result.ValueOrDie();
+    if (output_indices.empty()) {
+      *chunk_indices = nullptr;
+      *num_chunk_indices = 0;
+      return;
+    }
+
+    *chunk_indices = static_cast<int64_t*>(malloc(sizeof(int64_t) * output_indices.size()));
+    if (*chunk_indices) {
+      std::copy(output_indices.begin(), output_indices.end(), *chunk_indices);
+      *num_chunk_indices = output_indices.size();
+    } else {
+      *num_chunk_indices = 0;
+    }
+  } catch (...) {
+    *chunk_indices = nullptr;
```

Comment:
the exception(if throw) will be dropped. we need log the exception or pass the details message of error to the caller ...